python_scripts/linear_models_feature_engineering_classification.py

```diff
@@ -0,0 +1,437 @@
+# only one of the two features is greater than 0. The target class is 0
+# otherwise.
+
+# %%
+xor = pd.DataFrame(
+    np.random.RandomState(0).uniform(low=-1, high=1, size=(200, 2)),
+    columns=feature_names,
+)
+target_xor = np.logical_xor(xor["Feature #0"] > 0, xor["Feature #1"] > 0)
+target_xor = target_xor.astype(np.int32)
+xor["class"] = target_xor
+data_xor = xor[feature_names]
+
+# %% [markdown]
+#
+# We use matplotlib to visualized the data points in both datasets:
+
+# %%
+import matplotlib.pyplot as plt
+import seaborn as sns
+import warnings
+
+
+warnings.filterwarnings("ignore", module="seaborn")
+_, axs = plt.subplots(ncols=3, figsize=(14, 4))
+
+sns.scatterplot(
+    data=moons,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_moons,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[0],
+)
+sns.scatterplot(
+    data=gauss,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_gauss,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[1],
+)
+sns.scatterplot(
+    data=xor,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_xor,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[2],
+)
+axs[0].set_title("The moons dataset")
+_ = axs[1].set_title("The Gaussian quantiles dataset")
+_ = axs[2].set_title("The XOR dataset")
+
+
+# %% [markdown]
+#
+# We intuitively observe that for each dataset, the two classes are not
+# linearly separable as the classes fold around each other in a way no straight
+# line can separate the two classes.
+#
+# We can confirm this by fitting a linear model, such as logistic regression,
+# to each dataset and plot the decision boundary of the model. Let's first define
+# a helper function to plot the decision boundary of a model:
+
+# %%
+from sklearn.inspection import DecisionBoundaryDisplay
+
+
+def plot_decision_boundary(model):
+    _, axs = plt.subplots(ncols=3, figsize=(14, 4))
+
+    for ax, (data, target) in zip(
+        axs,
+        [
+            (data_moons, target_moons),
+            (data_gauss, target_gauss),
+            (data_xor, target_xor),
+        ],
+    ):
+        model.fit(data, target)
+        DecisionBoundaryDisplay.from_estimator(
+            model,
+            data,
+            response_method="predict_proba",
+            plot_method="pcolormesh",
+            cmap="RdBu",
+            alpha=0.8,
+            vmin=0,
+            vmax=1,
+            ax=ax,
+        )
+        DecisionBoundaryDisplay.from_estimator(
+            model,
+            data,
+            response_method="predict_proba",
+            plot_method="contour",
+            alpha=0.8,
+            levels=[0.5],
+            linestyles="--",
+            linewidths=2,
+            ax=ax,
+        )
+        sns.scatterplot(
+            data=data,
+            x=feature_names[0],
+            y=feature_names[1],
+            hue=target,
+            palette=["tab:red", "tab:blue"],
+            ax=ax,
+        )
+    return axs
+
+
+# %% [markdown]
+#
```

Comment:
```suggestion
```

Reply:
I insert those blank lines intentionally at the beginning of the markdown cells to make it easier to just press alt-q in VS Code to rewrap the markdown content of the cell without having to select only the lines I want.

If I don't do this, re-wrapping by pressing alt-q on this cell would yield the following invalid markdown:

```
# %% [markdown] Now let's define our logistic regression model and plot its decision
# boundary on the three datasets:
```

Reply:
Also note: the extra blank space is not visible for the end-user, so we don't need to be consistent everywhere. It's just a convenience when editing with VS Code.